src/persist/src/location.rs

```diff
@@ -285,6 +320,34 @@ pub struct VersionedData {
     pub data: Vec<u8>,
 }
 
+impl<T: Codec> From<(SeqNo, &T)> for VersionedData {
+    fn from(x: (SeqNo, &T)) -> Self {
+        let (seqno, t) = x;
+        let mut ret = VersionedData {
+            seqno,
+            data: Vec::new(),
+        };
+        Codec::encode(t, &mut ret.data);
+        ret
+    }
+}
+
+impl<T: Codec> TryFrom<&VersionedData> for (SeqNo, T) {
+    type Error = ExternalError;
+
+    fn try_from(x: &VersionedData) -> Result<Self, Self::Error> {
+        let t = T::decode(&x.data).map_err(|err| {
+            ExternalError::from(anyhow!(
```

Comment:
Is it an `ExternalError` because some system might have corrupted it. To me, external errors are when some system that we use messes up, but this seems more like an internal problem. Just another sign that our error naming/design is still not where it should be, I guess... 😅

Reply:
I don't feel strongly and I def agree that our error story needs work, but I did actually document corruption in the ExternalError rustdoc :)

```
/// An error coming from an underlying durability system (e.g. s3) or from
/// invalid data received from one.
```

Reply:
👍